### PR TITLE
[WIP] Refactored markdown plugin, added some templates

### PR DESCRIPTION
**Pull request by grgprarup**
This PR includes the following changes:
- Refactored the existing markdown plugin, so that it works for markdown file metadata
- Added some sample templates for different slides
- Mustache package is used for template rendering

Note:: Ignore changes in `markdown.esm.js` and `markdown.js`, those files are changed with the new build.

### CODE REVIEW

File: plugin/markdown/plugin.js [hunk 1]
```diff
@@ -174,25 +184,38 @@ const Plugin = () => {
...
-		var markdownSections = '';
+		let markdownSections = '';
 
 		// flatten the hierarchical stack, and insert <section data-markdown> tags
-		for( var i = 0, len = sectionStack.length; i < len; i++ ) {
+		for( let i = 0, len = sectionStack.length; i < len; i++ ) {
+			let classMetadata = '';
+			if (parsedMetadata) {
+				options.metadata = parsedMetadata.slides[i];
+				classMetadata = ' class=' + options.metadata.slideType;
```
Comment (saw-jan) [inline plugin/markdown/plugin.js]: if that works
```suggestion
			if (parsedMetadata) {
				options.metadata = parsedMetadata.slides[i];
				options.attributes += ' class=' + options.metadata.slideType;
```

File: plugin/markdown/plugin.js [hunk 2]
```diff
@@ -209,7 +232,7 @@ const Plugin = () => {
 
 		return new Promise( function( resolve ) {
 
-			var externalPromises = [];
+            const externalPromises = [];
```
Comment (saw-jan) [inline plugin/markdown/plugin.js]:
```suggestion
                          const externalPromises = [];
```

File: plugin/markdown/plugin.js (hunk 1 above)
Comment (grgprarup) [reply plugin/markdown/plugin.js]: Nice catch!!!


:smiley:     ***Reviewing Code, Improves Its Quality***

File: index.html [hunk 3]
```diff
@@ -12,19 +12,34 @@
...
+				<section data-markdown="examples/layout-template.md">
+				</section>
 			</div>
 		</div>
 
 		<script src="dist/reveal.js"></script>
 		<script src="plugin/notes/notes.js"></script>
 		<script src="plugin/markdown/markdown.js"></script>
 		<script src="plugin/highlight/highlight.js"></script>
+		<script src="https://unpkg.com/mustache@latest"></script>
```
Comment (individual-it) [inline index.html]: mustache is installed by the package.json, so it should not be loaded from the external source

File: plugin/markdown/plugin.js [hunk 4]
```diff
@@ -1,10 +1,12 @@
 /*!
- * The reveal.js markdown plugin. Handles parsing of
+ * The reveal.js Markdown plugin. Handles parsing of
  * markdown inside of presentations as well as loading
- * of external markdown documents.
+ * of external Markdown documents.
```
Comment (individual-it) [inline plugin/markdown/plugin.js]: we should not have those changes in this PR

File: plugin/markdown/plugin.js [hunk 5]
```diff
@@ -38,15 +40,15 @@ const Plugin = () => {
 	function getMarkdownFromSlide( section ) {
 
 		// look for a <script> or <textarea data-template> wrapper
-		var template = section.querySelector( '[data-template]' ) || section.querySelector( 'script' );
+		const template = section.querySelector( '[data-template]' ) || section.querySelector( 'script' );
```
Comment (individual-it) [inline plugin/markdown/plugin.js]: same here, can we reduce the changes to what is really needed in this PR?

**Review by saw-jan — COMMENTED** [review]
Looks good for me. But I don't know what is just reviewed :sweat_smile:

File: index.html [hunk 6]
```diff
@@ -12,19 +12,34 @@
...
+			.front-page h1 {
+				color: #000000;
+			}
+			.second-page {
+				background-color: #6f4545;
+				background-size: cover;
+				background-position: center;
+				background-repeat: no-repeat;
+			}
+		</style>
```
Comment (individual-it) [inline index.html]: can those things be part of the template file?

File: plugin/markdown/plugin.js [hunk 7]
```diff
@@ -174,13 +184,25 @@ const Plugin = () => {
 			wasHorizontal = isHorizontal;
 		}
 
+		if (metadata){
+			try {
+				parsedMetadata = yaml.load(metadata);
+			} catch (error) {
+				console.error("Error while parsing metadata", error)
```
Comment (individual-it) [inline plugin/markdown/plugin.js]: this should be somehow displayed to the user who creates the presentation, maybe show it actually on the first slide

File: plugin/markdown/plugin.js [hunk 8]
```diff
@@ -136,19 +141,22 @@ const Plugin = () => {
...
 
-		var matches,
+		let matches,
 			lastIndex = 0,
 			isHorizontal,
 			wasHorizontal = true,
 			content,
-			sectionStack = [];
+			sectionStack = [],
+			isMetadata = true,
```
Comment (individual-it) [inline plugin/markdown/plugin.js]: so do we always expect metadata? What if the first slide is actually just a slide?

**Review by individual-it — CHANGES_REQUESTED** [review]

File: index.html (hunk 6 above)
Comment (grgprarup) [reply index.html]: Yeah. That will go in the template file.

**Review by saw-jan — COMMENTED** [review]
file `utils/exporter.js` is empty

File: plugin/markdown/plugin.js (hunk 8 above)
Comment (grgprarup) [reply plugin/markdown/plugin.js]: The logic changed for this. I think it will work now.

File: js/reveal.js [hunk 9]
```diff
@@ -2887,7 +2887,7 @@ export default function( revealElement, options ) {
 		showPreview,
 		hidePreview: closeOverlay,
 
-		// Adds or removes all internal event listeners
+		// Adds or removes all internal event listengit reers
```
Comment (individual-it) [inline js/reveal.js:2890]: typo?

**Pull request closed without merging**